### PR TITLE
Add `__len__` method to IndexBatchSamplerWrapper

### LINKED ISSUE #7664: Trainer.predict is not working on TPUs
## 🐛 Bug

Trainer.predict does not seem to run correctly when using TPU.
 I get an error "TPUs do not currently support IterableDataset objects, the dataset must implement `__len__`. HINT: You can mock the length on your dataset to bypass this MisconfigurationException." even though I am using a dataset that implements `__len__`.

### To Reproduce

If you create an instance of Trainer with tpu_cores=1 or 8, and run predict, an error occurs.

Notebook: https://colab.research.google.com/drive/1yHI-eKlfdq_zXUSPWmNvivEPJKl8RZbA?usp=sharing

### Expected behavior

Return the output of the model.

### Environment

 - PyTorch Version (e.g., 1.0): 1.8
 - Any other relevant information: Google Colab with TPU accelerator

### PR DESCRIPTION
**Pull request by kaushikb11**
## What does this PR do?

Fixes #7664 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/models/test_tpu.py [hunk 1]
```diff
@@ -495,3 +495,22 @@ def teardown(self, stage):
 
     model = DebugModel()
     tpipes.run_model_test(trainer_options, model, on_gpu=False, with_hpc=False)
+
+
+@RunIf(tpu=True)
+@pl_multi_process_test
+def test_predict_tpu_multi_cores(tmpdir):
```
Comment (awaelchli) [inline tests/models/test_tpu.py:502]: this test makes no assertions, which means it will pass if the trainer does not train or the prediction loop does not return any predictions

**Review by awaelchli — APPROVED** [review]

File: tests/models/test_tpu.py (hunk 1 above)
Comment (kaushikb11) [reply tests/models/test_tpu.py:502]: Added an assertion for trainer state.

File: tests/models/test_tpu.py [hunk 2]
```diff
@@ -495,3 +495,24 @@ def teardown(self, stage):
...
+    model = BoringModel()
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        progress_bar_refresh_rate=0,
+        max_epochs=2,
+        limit_train_batches=2,
+        limit_val_batches=2,
+        tpu_cores=8,
+    )
+    trainer.fit(model)
```
Comment (carmocca) [inline tests/models/test_tpu.py:515]: Does the test have to call `fit` before to ensure `predict` works?

File: tests/models/test_tpu.py [hunk 3]
```diff
@@ -495,3 +495,24 @@ def teardown(self, stage):
...
+    trainer.predict(model, DataLoader(RandomDataset(32, 2000), batch_size=32))
```
Comment (carmocca) [inline tests/models/test_tpu.py:516]: Do we need the 2k here for the test?

File: tests/models/test_tpu.py [hunk 4]
```diff
@@ -495,3 +495,24 @@ def teardown(self, stage):
 
     model = DebugModel()
     tpipes.run_model_test(trainer_options, model, on_gpu=False, with_hpc=False)
+
+
+@RunIf(tpu=True)
+@pl_multi_process_test
+def test_predict_tpu_multi_cores(tmpdir):
```
Comment (carmocca) [inline tests/models/test_tpu.py:502]: I think an unit test would be better suited for this PR

Comment (kaushikb11) [reply tests/models/test_tpu.py:502]: What do you have in mind? Just check the `len` attribute for `IndexBatchSamplerWrapper`?

Comment (carmocca) [reply tests/models/test_tpu.py:502]: Yeah, simple and effective 😜

Comment (kaushikb11) [reply tests/models/test_tpu.py]: Done!

**Review by carmocca — APPROVED** [review]
